main/src/ca/uwaterloo/flix/language/phase/jvm/AsmOps.scala

```diff
@@ -120,19 +122,50 @@ object AsmOps {
   }
 
   /**
-    * Returns the array type code for the value of the type specified by `tpe`
+    * Returns the array store instruction for arrays of the given JvmType tpe
     */
-  def getArrayTypeCode(tpe: JvmType): Int = tpe match {
-    case JvmType.Void => throw InternalCompilerException(s"Unexpected type $tpe", SourceLocation.Unknown)
-    case JvmType.PrimBool => T_BOOLEAN
-    case JvmType.PrimChar => T_CHAR
-    case JvmType.PrimFloat => T_FLOAT
-    case JvmType.PrimDouble => T_DOUBLE
-    case JvmType.PrimByte => T_BYTE
-    case JvmType.PrimShort => T_SHORT
-    case JvmType.PrimInt => T_INT
-    case JvmType.PrimLong => T_LONG
-    case JvmType.Reference(_) => throw InternalCompilerException(s"Expected primitive type. Actual type: $tpe", SourceLocation.Unknown)
+  def getArrayStoreInstruction(tpe: BackendType): Int = tpe match {
+    case BackendType.Bool => BASTORE
+    case BackendType.Char => CASTORE
+    case BackendType.Int8 => BASTORE
+    case BackendType.Int16 => SASTORE
+    case BackendType.Int32 => IASTORE
+    case BackendType.Int64 => LASTORE
+    case BackendType.Float32 => FASTORE
+    case BackendType.Float64 => DASTORE
+    case BackendType.Reference(_) => AASTORE
+    case BackendType.Array(_) => AASTORE
+  }
+
+  def getArrayTypeCode(tpe: BackendType): Option[Int] = tpe match {
+    case BackendType.Bool => Some(T_BOOLEAN)
+    case BackendType.Char => Some(T_CHAR)
+    case BackendType.Int8 => Some(T_BYTE)
+    case BackendType.Int16 => Some(T_SHORT)
+    case BackendType.Int32 => Some(T_INT)
+    case BackendType.Int64 => Some(T_LONG)
+    case BackendType.Float32 => Some(T_FLOAT)
+    case BackendType.Float64 => Some(T_DOUBLE)
+    case BackendType.Array(_) | BackendType.Reference(_) => None
+  }
+
+  def getArrayType(tpe: BackendType): String = {
+    @tailrec
+    def visit(tt: BackendType, acc: String): String = tt match {
+      case BackendType.Bool => acc + "Z"
+      case BackendType.Char => acc + "C"
+      case BackendType.Int8 => acc + "B"
+      case BackendType.Int16 => acc + "S"
+      case BackendType.Int32 => acc + "I"
+      case BackendType.Int64 => acc + "J"
+      case BackendType.Float32 => acc + "F"
+      case BackendType.Float64 => acc + "D"
+      case BackendType.Reference(ref) if acc.isEmpty => ref.jvmName.toInternalName
```

Comment:
Other than this case, this function is the same as calling `tt.toDescriptor`, why is that?

Reply:
I see. For some reason the jvm is quite inconsistent when it comes to arrays of objects; if the component type is a non-array object then it uses the internal name, otherwise the object must use its descriptor with a `[` prepended.

Reply:
I refactor `BackendType.Array` to override the `toDescriptor` method